[DOC-2597] Revise - Add an Azure Cloud connector (#715)
harness/developer-hub: docs/platform/7_Connectors/add-a-microsoft-azure-connector.md

---
title: Add a Microsoft Azure Cloud Provider connector
description: The Azure Cloud Provider connector connects Harness to Azure.
# sidebar_position: 2
helpdocs_topic_id: 9epdx5m9ae
helpdocs_category_id: o1zhrfo8n5
helpdocs_is_private: false
helpdocs_is_published: true
---

> With the Microsoft Azure Cloud connector, your Harness pipelines can pull Azure artifacts and deploy your applications to Azure.

Reviewer: ```suggestion
With the Microsoft Azure Cloud connector, your Harness pipelines can pull Azure artifacts, provision Azure infrastructure, and deploy your applications to Azure.
```

Author: change. Also changed title and added cross-link to Azure Repos connector topic, as discussed.

> Currently, the Microsoft Azure Cloud Connector is for ACR, AKS, Web Apps, and Virtual Machines for Traditional (SSH) deployments. Support for other services such as ARM and Blueprint are coming soon.

Reviewer 2: We do support ARM and Blueprint in the execution steps section, so I would remove this last sentence.

Author: Removed "coming soon" sentence and added Blueprints and ARM to the compatibilities list.

Using Harness **Cloud Cost Management (CCM)**? See [Set Up Cloud Cost Management for Azure](../../cloud-cost-management/1-onboard-with-cloud-cost-management/set-up-cloud-cost-management/set-up-cost-visibility-for-azure.md).
The Microsoft Azure Cloud Connector is for ACR, AKS, ARM, Blueprints, Web Apps, and Virtual Machines for Traditional (SSH) deployments.

### Before you begin
:::tip

* [Learn Harness' Key Concepts](../../getting-started/learn-harness-key-concepts.md)
* [CD Pipeline Basics](../../continuous-delivery/onboard-cd/cd-concepts/cd-pipeline-basics.md)

### Limitations

* Currently, the Microsoft Azure Cloud Connector is for: ACR, AKS, Web Apps, and Virtual Machines for Traditional (SSH) deployments. Support for other services such as ARM and Blueprint are coming soon.
* To use an AKS cluster for deployment, the AKS cluster must have local accounts enabled (AKS property `disableLocalAccounts=false`).
If you're using Harness **Cloud Cost Management (CCM)**, you can [Set Up Cloud Cost Management for Azure](../../cloud-cost-management/1-onboard-with-cloud-cost-management/set-up-cloud-cost-management/set-up-cost-visibility-for-azure.md).

### Visual Summary
:::

> The following example shows how to connect Harness to Azure using the Azure Cloud Connector and an Azure App registration.

Author: moved it down to somewhere where the actual image is more relevant.

> Harness Azure Cloud connectors that you'll use to connect to ACR must have the **Reader** role, at minimum, or an equivalent or better custom role.

Reviewer: ```suggestion
The Harness Azure Cloud connectors that you'll use to connect Harness to ACR must have the **Reader** role, at minimum. You can also use a custom role that includes the permissions of the **Reader** role.
```

Author: fixed

```mdx-code-block
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
```

```mdx-code-block
<Tabs>
  <TabItem value="reader" label="Reader" default>
```

> The **Reader** role must be assigned at the **Subscription** or **Resource Group** level that is used by the Application (Client) Id that you'll use in the Azure Cloud connector's settings. The application must have permission to list **all** container registries.

Reviewer: I feel like we're using `Azure Cloud connector` and `Azure Cloud Provider connector` interchangeably. 

I'd be fine with `Azure connector`.

Author: changed to Azure connector throughout, except where it specifically referenced the UI.

                    "Microsoft.ContainerService/managedClusters/pods/write",
                    "Microsoft.ContainerService/managedClusters/pods/delete",
                    "Microsoft.ContainerService/managedClusters/services/read",
                    "Microsoft.ContainerService/managedClusters/services/write",
                    "Microsoft.ContainerService/managedClusters/services/delete",
                    "Microsoft.ContainerService/managedClusters/apps/statefulsets/read",
                    "Microsoft.ContainerService/managedClusters/apps/statefulsets/write",
                    "Microsoft.ContainerService/managedClusters/apps/statefulsets/delete",
                    "Microsoft.ContainerService/managedClusters/apps/replicasets/read",
                    "Microsoft.ContainerService/managedClusters/apps/replicasets/write",
                    "Microsoft.ContainerService/managedClusters/apps/replicasets/delete"
                ],
                "notDataActions": []
            }
        ]
    }
}
```

```mdx-code-block
  </TabItem2>
</Tabs2>
```

## Connect Harness to Azure Kubernetes Services (AKS)

There are three options for connecting Harness to an AKS cluster:

> * Use the platform-agnostic [Kubernetes cluster connector](add-a-kubernetes-cluster-connector.md). with a Kubernetes delegate.

Reviewer: ```suggestion
* Use the platform-agnostic [Kubernetes cluster connector](add-a-kubernetes-cluster-connector.md) with a Kubernetes or Helm delegate.
```

* Use a **Microsoft Azure Cloud Provider connector**, as described in this topic, with a Kubernetes delegate.
    + You'll need to install a [Kubernetes delegate](../2_Delegates/advanced-installation/install-a-kubernetes-delegate.md) in the target AKS cluster, and then use the delegate's credentials for the Azure Cloud connector's authentication method.
	+ You'll need to provide the Microsoft Azure Environment.
	+ If you use a User Assigned Managed Identity, you'll need to provide the Application (client) Id.
	+ If you use a System Assigned Managed Identity, you won't need to provide any Ids.
* Use a **Microsoft Azure Cloud Connector** with Service Principal or Managed Identity credentials, as described in this topic.
    + You must assign the **Owner** role or an equivalent custom role, as explained in [AKS role requirements](#aks-role-requirements).

### AKS cluster setup requirements

* AKS managed AAD, enabled or disabled.
* Kubernetes RBAC, enabled.
* Azure RBAC, enabled or disabled.
	+ For more information, go to the **Deployments (CD)** section of the [Kubernetes cluster connector settings reference](ref-cloud-providers/kubernetes-cluster-connector-settings-reference.md).

> * AKS property `disableLocalAccounts`, enabled or disabled, unless used for deployment.

Reviewer 2: Remove this part regarding 'disableLocalAccounts' property. It is not valid anymore.

>     + To use an AKS cluster for deployment, the AKS cluster must have local accounts enabled (`disableLocalAccounts=false`).

Reviewer 2: Remove this part as well.

Author: removed

   For more information about managed identities, go to the follow Azure documentation:

   * [Use managed identities in Azure Kubernetes Service](https://docs.microsoft.com/en-us/azure/aks/use-managed-identity)
   * [How to use managed identities with Azure Container Instances](https://docs.microsoft.com/en-us/azure/container-instances/container-instances-managed-identity) from Azure.

   ![](./static/add-a-microsoft-azure-connector-66.png)

3. If you selected **User Assigned Managed Identity**, input the Managed Identity's **Client Id**, which you can find in your Azure **Managed Identities**. You can also use a [Pod Assigned Managed identity](https://docs.microsoft.com/en-us/azure/aks/use-azure-ad-pod-identity).

   ![](./static/add-a-microsoft-azure-connector-67.png)

4. Select **Continue**

</details>

## Select connectivity mode

1. Select how you want Harness to connect to Azure:

>    * **Connect through Harness Platform:** Use a direct, secure communication between Harness and Azuree.

Reviewer: ```suggestion
   * **Connect through Harness Platform:** Use a direct, secure communication between Harness and Azure.
```

> 1. Select how you want Harness to connect to Azure:

Reviewer 2: Important to mention here is that if the user chooses User or System Assigned Managed Identity connector then it must use "Connect through a Harness delegate" connectivity mode.

Author: Because User Assigned Managed Identity and System Assigned Managed Identity are only available if you choose "Use the credentials of a specific Harness delegate", I included the statement: "You must choose this option (Connect through a Harness Delegate) if you chose to inherit delegate credentials."

It is not quite obvious from the raw markdown, but there are two expandable sections describing the settings associated with "Specify credentials here" and "Inherit credentials from the delegate". The explanations for User/System Assigned Managed Identity are only revealed if you expand the "Inherit credentials from the delegate" section.

After I commit my changes based on your comments, the preview build should regenerate a new preview build link so you can see how the page will look when published.

Author: Here are some screenshots showing the expandable sections: 
![configure credentials section](https://user-images.githubusercontent.com/36110273/220437849-dbfaa3c7-48b3-44e1-87b6-d032c3eb3c0b.png)

![configure credentials section with inherit from delegate expanded](https://user-images.githubusercontent.com/36110273/220438014-0ba6fdf6-9d8c-44b9-91cf-7e9b77a5450a.png)

   * **Connect through a Harness Delegate:** Harness communicates with Azure through a Harness delegate in your Azure subscription or AKS cluster. You must choose this option if you chose to inherit delegate credentials.

2. If connecting through a Harness delegate, select either:

>    * **Use any available Delegate:** Harness selects an available Delegate at runtime.

Reviewer: ```suggestion
   * **Use any available Delegate:** Harness selects an available delegate at runtime.
```

3. Select **Save and Continue** to run the connection test, and then, if the test succeeds, select **Finish**. The connection test confirms that your authentication and delegate selections are valid.

   If the connection test fails, make sure that your delegate is running and that your credentials are valid. For example, check that the secret has not expired in your App registration.

## Using ${HARNESS\_KUBE\_CONFIG\_PATH} with Azure

The Harness `${HARNESS_KUBE_CONFIG_PATH}` expression resolves to the path to a Harness-generated kubeconfig file containing the credentials you provided to Harness.

The credentials can be used by kubectl commands by exporting its value to the `KUBECONFIG` environment variable.

For example, you could use this shell script in a Harness Run step:

```
export KUBECONFIG=${HARNESS_KUBE_CONFIG_PATH} kubectl get pods -n <namespace>
```

Steps can be executed on any delegate or you can select specific delegates using the [Delegate Selector](../2_Delegates/manage-delegates/select-delegates-with-selectors.md) setting.

For Azure deployments, note the following:

> * If the Azure connector used in the stage's **Infrastructure** uses Azure Managed Identity for authentication, then the Run step with the shell script must use a **Delegate Selector** for a delegate running in AKS.

Reviewer: ```suggestion
* If the Azure connector used in the stage's infrastructure uses Azure Managed Identity for authentication, then the Shell Script step must use a **Delegate Selector** for a delegate running in AKS.
```

Reviewer: there is no Run step in CD. Run is in CI. CD has Shell Script.

Author: Got it, thanks!

> * If the Azure connector used in the stage's **Infrastructure** uses Azure Service Principal for authentication, then the Run step with the shell script can use any Delegate.

Reviewer: ```suggestion
* If the Azure connector used in the stage's infrastructure uses Azure Service Principal for authentication, then the Shell Script step can use any Delegate.
```
